lib/procore/services/batch_sync.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+# Used for large amounts of data syncing to Procore at one time
+# This class will batch the records
+module Procore
+  module Services
+    class BatchSync
+      BATCH_SIZE = 500.freeze
+
+      def initialize(url:, arguments: {}, updates:, connection:, batch_size: BATCH_SIZE)
+        @url = url
+        @arguments = arguments
+        @updates = updates
+        @connection = connection
+        @batch_size = batch_size
+      end
+
+      def execute
+        entities = []
+        errors = []
+        batches.each do |batch|
+          sync_arguments = arguments.merge(updates: batch.compact)
+          response = connection.patch(url, sync_arguments).body
+          entities += response['entities'] if response['entities']&.is_a?(Array)
+          errors += response['errors'] if response['errors']&.is_a?(Array)
+        end
+
+        { entities: entities, errors: errors }
+      end
```

Comment:
NB: I think we could also change this block to a better enumerable like a reduce or each_with_object.

I'm thinking something like:

```
def execute
  batches.each_with_object({entities:[],errors:[]}) do |batch, results|
    sync_arguments = arguments.merge(updates: batch.compact)
    response = connection.patch(url, sync_arguments).body
    results[:entities] += response['entities'] if response['entities']&.is_a?(Array)
    results[:errors] += response['errors'] if response['errors']&.is_a?(Array)
  end
end
```

What do you think?

Reply:
I like @AllPurposeName idea of using `each_with_object`